### PR TITLE
#1 Integration with Lagom Java DSL

### PR DESCRIPTION
**Pull request by ihostage**
- [x] Initial implementation
- [x] Tests
- [x] Javadoc

### CODE REVIEW

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Fluent API for creating authorizers in implementation Lagom services.
+ *
+ * @since 1.0.0
+ */
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: Done

File: shared/src/main/java/org/pac4j/lagom/javadsl/LagomWebContext.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.Collection;
+import java.util.Map;
+
+import static java.lang.String.format;
+
+/**
+ * <p>Implementation web context of PAC4J for Lagom framework.</p>
+ * <p>Context is immutable and the {@link SessionStore} is not supported.</p>
+ *
+ * @since 1.0.0
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/javadsl/LagomWebContext.java]: Done

File: shared/src/main/java/org/pac4j/lagom/javadsl/LagomWebContext.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+ */
+public class LagomWebContext implements WebContext {
+
+    private RequestHeader requestHeader;
+
+    public LagomWebContext(RequestHeader requestHeader) {
+        this.requestHeader = requestHeader;
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/javadsl/LagomWebContext.java]: You right, I removed `{@inheritDoc}`

File: shared/src/test/java/org/pac4j/lagom/javadsl/Clients.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package org.pac4j.lagom.javadsl;
+
+interface Clients {
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/test/java/org/pac4j/lagom/javadsl/Clients.java]: >We already have a class called Clients in pac4j-core: I'm not sure this is a good idea to have the same name here.

This interface using only in test classpath, it shouldn't be a big problem 😄 

> What's this class for? Can you add some Javadoc?

It's just place for names of clients, that will be used for various tests.

>Maybe ClientNames, but I think we can live without it...

I renamed this interface.

File: shared/src/test/java/org/pac4j/lagom/javadsl/TestModule.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import org.pac4j.http.client.direct.HeaderClient;
+
+import javax.inject.Named;
+
+import static org.pac4j.core.context.HttpConstants.AUTHORIZATION_HEADER;
+import static org.pac4j.lagom.Authorizers.anonymous;
+import static org.pac4j.lagom.Authorizers.authenticated;
+import static org.pac4j.lagom.javadsl.Clients.HEADER_CLIENT;
+
+public class TestModule extends AbstractModule implements ServiceGuiceSupport {
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/test/java/org/pac4j/lagom/javadsl/TestModule.java]: Done.

PS: Sorry, I don't usually write javadoc in tests 😄

File: shared/src/test/java/org/pac4j/lagom/javadsl/TestService.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+import akka.NotUsed;
+import com.lightbend.lagom.javadsl.api.Descriptor;
+import com.lightbend.lagom.javadsl.api.Service;
+import com.lightbend.lagom.javadsl.api.ServiceCall;
+
+import static com.lightbend.lagom.javadsl.api.Service.named;
+import static com.lightbend.lagom.javadsl.api.Service.pathCall;
+
+public interface TestService extends Service {
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/test/java/org/pac4j/lagom/javadsl/TestService.java]: Done

File: shared/src/test/java/org/pac4j/lagom/javadsl/TestServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import org.pac4j.core.config.Config;
+
+import javax.inject.Inject;
+
+import static org.pac4j.lagom.Authorizers.authenticated;
+import static org.pac4j.lagom.javadsl.Clients.HEADER_CLIENT;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class TestServiceImpl implements TestService, SecuredService {
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/test/java/org/pac4j/lagom/javadsl/TestServiceImpl.java]: Done

File: shared/src/test/java/org/pac4j/lagom/javadsl/TestModule.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Provides
+    @Named(HEADER_CLIENT)
+    protected HeaderClient provideHttpClient() {
+        HeaderClient headerClient = new HeaderClient(AUTHORIZATION_HEADER, (credentials, webContext) -> {
+            final CommonProfile profile = new CommonProfile();
+            profile.setId(((TokenCredentials)credentials).getToken());
+            credentials.setUserProfile(profile);
+        });
+        headerClient.setName(HEADER_CLIENT);
```
[the remark this answers is not in the document]
Comment (ihostage) [reply shared/src/test/java/org/pac4j/lagom/javadsl/TestModule.java]: I planing write tests for header client with jwt authenticator. (issue #6)
And I decided set the custom name for all clients in tests.

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    public static <U extends CommonProfile> Authorizer<U> authenticated() {
+        return new IsAuthenticatedAuthorizer<>();
+    }
+
+    public static <U extends CommonProfile> Authorizer<U> anonymous() {
+        return new IsAnonymousAuthorizer<>();
+    }
+
+    public static <U extends CommonProfile> Authorizer<U> anyRole(String ... roles) {
```
Comment (leleuj) [inline shared/src/main/java/org/pac4j/lagom/Authorizers.java]: About the naming, I think I'd prefer `requireAnyRole` as this is the name of the `Authorizer` and the name we generally use (like the `@RequireAnyRole` annotation in the `spring-webmvc-pac4j` library).

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        return new RequireAnyRoleAuthorizer<>(roles);
+    }
+
+    public static <U extends CommonProfile> Authorizer<U> anyRole(List<String> roles) {
```
Comment (leleuj) [inline shared/src/main/java/org/pac4j/lagom/Authorizers.java]: `requireAnyRole`

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        return new RequireAnyRoleAuthorizer<>(roles);
+    }
+
+    public static <U extends CommonProfile> Authorizer<U> anyRole(Set<String> roles) {
```
Comment (leleuj) [inline shared/src/main/java/org/pac4j/lagom/Authorizers.java]: `requireAnyRole`. I don't see the `requireAllRoles` even if I guess that the `and` handles this use case...

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+    public static <U extends CommonProfile> OrAuthorizer<U> or(Authorizer<U>... authorizers) {
+        return new OrAuthorizer<>(Arrays.asList(authorizers));
+    }
+
+    /**
+     * The conjunction of authorizers.
+     * @param <U> Type of profile
+     */
+    private static class AndAuthorizer<U extends CommonProfile> implements Authorizer<U> {
```
Comment (leleuj) [inline shared/src/main/java/org/pac4j/lagom/Authorizers.java]: I think this should into the `pac4j-core` project.

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            return true;
+        }
+
+    }
+
+    /**
+     * The disjunction of authorizers.
+     * @param <U> Type of profile
+     */
+    private static class OrAuthorizer<U extends CommonProfile> implements Authorizer<U> {
```
Comment (leleuj) [inline shared/src/main/java/org/pac4j/lagom/Authorizers.java]: I think this should into the `pac4j-core` project.

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java (hunk 9 above)
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: Done

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java (hunk 11 above)
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: Done

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java (hunk 13 above)
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: I created PR https://github.com/pac4j/pac4j/pull/1199
1. Move code to `pac4j-core`.
2. Mark this class how `deprecated`. (until release 3.4.0)

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java (hunk 12 above)
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: Done https://github.com/pac4j/pac4j/pull/1199

File: shared/src/main/java/org/pac4j/lagom/Authorizers.java (hunk 13 above)
Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: 👌 I will remove this class and migrate tests after merging https://github.com/pac4j/pac4j/pull/1199

Comment (ihostage) [reply shared/src/main/java/org/pac4j/lagom/Authorizers.java]: Done

**Comment by leleuj** [comment]
We're good to go. Merging...

**Comment by ihostage** [comment]
🎉

**Pull request merged**